### PR TITLE
Fixes 79: Search and filter content sources list

### PR DESCRIPTION
**Pull request by rverdile**
**Adds:** 
- `ContentListFilters.tsx`
- `ContentListChips.tsx`
- `DropdownSelect.tsx`
- ~~`ContentListToolbar.tsx`~~

ContentListFilters, ContentListChips, and DropdownSelect all work nicely together with the existing table to make the search bar with filters. This is what I had originally (and what ContentListTable currently renders).

 ~~_Then_, I found the [patternfly Toolbar component](https://www.patternfly.org/v4/components/toolbar/), which is maybe what I should've used, so I tried implementing the same thing with that. But, that implementation is still buggy and seems a lot messier than the first one.~~ Not using this.

Still needs
- [x] ~~Fix ContentListToolbar?~~
- [x] API call to repository_params endpoint (right now uses mocked data)

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMSCONTENT-79

File: src/components/ContentListTable/ContentListTable.tsx [hunk 1]
```diff
@@ -131,6 +181,17 @@ const ContentListTable = () => {
...
       </Flex>
+      <Flex className={classes.chips}>
+        <ContentListChips
+            searchQuery={searchQuery}
+            debouncedSearchQuery={debouncedSearchQuery}
+            setSearchQuery={setSearchQuery}
+            selectedVersions={selectedVersions}
+            setSelectedVersions={setSelectedVersions}
+            selectedArches={selectedArchitectures}
+            setSelectedArches={setSelectedArchitectures}
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:192]: Make these variables internal state using react context within this component.
Do not pass down setters/variables to children, use react context to get them within the child components instead. 
I did something similar in katello's bulkDeleteModal, but the idea is that you only reveal the logic needed for the parent component, in this case the setSearchQuery variable should be all you need.

File: src/components/ContentListToolbar/ContentListToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    const [selectedFilters, setSelectedFilters] = useState({
+        filterType: 'Name/URL',
+        versions: [],
+        architectures: [],
+    });
+    const [filterIsOpen, setFilterIsOpen] = useState(false);
+    const [versionIsOpen, setVersionIsOpen] = useState(false);
+    const [archIsOpen, setArchIsOpen] = useState(false);
+
+    const selectFrom = (options: Array<string>) => {
```
Comment (Andrewgdewar) [inline src/components/ContentListToolbar/ContentListToolbar.tsx:57]: where possible, for this and all other simple functions, use a direct return. 
someFunc = (vars) => vars.dothings()

File: src/components/ContentListTable/ContentListChips.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    setSelectedArches,
+}: ContentListChipsProps) => {
+
+    const deleteItem = (id: string, chips, setChips) => {
+        const copyOfChips = [...chips];
+        const filteredCopy = copyOfChips.filter(chip => chip !== id);
+        setChips(filteredCopy);
+    };
+
+    const onClick = () => {
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListChips.tsx:30]: A more meaningful name here would be clearFilters

**Comment by rverdile** [comment]
You should know, on the backend main branch, there's a bug with filtering list by URL.

URL moved from `repository_configurations` to `repostories`, which causes a "URL column not found" error. A temporary work around is to remove the URL searching from this line and just search by name.

https://github.com/content-services/content-sources-backend/blob/main/pkg/dao/repository.go#L93

File: src/components/ContentListTable/ContentListTable.tsx [hunk 4]
```diff
@@ -8,35 +8,66 @@ import {
...
+]
+}
 
 const useStyles = createUseStyles({
   actionContainer: {
     backgroundColor: global_BackgroundColor_100.value,
     justifyContent: 'space-between',
     padding: '16px 24px',
   },
+  chips: {
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:66]: Maybe rename to chipsContainer**

File: src/components/ContentListTable/ContentListTable.tsx [hunk 5]
```diff
@@ -51,15 +82,37 @@ const ContentListTable = () => {
   const [page, setPage] = useState(1);
   const [perPage, setPerPage] = useState(storedPerPage);
 
+  const context = useContext(ContentListContext)
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:85]: de-structure here.

File: src/components/ContentListTable/ContentListTable.tsx [hunk 6]
```diff
@@ -51,15 +46,43 @@ const ContentListTable = () => {
...
+    names.forEach(name => {
+          const found = namesLabels.find(v => v.name === name )
+          if(found) {
+            labels.push(found.label)
+          }
+        }
+    );
+    return labels
+  };
+
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:70]: Could this be done in context? As the labels are consumed here, the logic repoParams and versions/arches can take place in context as well.

File: src/components/ContentListTable/ContentListTable.tsx [hunk 7]
```diff
@@ -108,82 +131,80 @@ const ContentListTable = () => {
...
-            variant={PaginationVariant.bottom}
-            onPerPageSelect={onPerPageSelect}
-          />
-        </FlexItem>
-      </Flex>
-    </Grid>
+      <Grid>
+        <Flex className={classes.actionContainer}>
+          <FlexItem>
+              <ContentListFilters/>
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:137]: 👍

File: src/components/ContentListTable/ContentListTable.tsx [hunk 8]
```diff
@@ -108,82 +131,80 @@ const ContentListTable = () => {
...
+              page={page}
+              onSetPage={onSetPage}
+              widgetId='pagination-options-menu-top'
+              isCompact
+              onPerPageSelect={onPerPageSelect}
+            />
+          </FlexItem>
+        </Flex>
+        <Flex className={classes.chipsContainer}>
+          <ContentListChips/>
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:153]: 👍

File: src/components/ContentListTable/ContentListFilters.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import DropdownSelect from '../DropdownSelect/DropdownSelect';
+import { FilterIcon } from '@patternfly/react-icons';
+import { ContentListContext } from './ContentListContext';
+
+const ContentListFilters = () => {
+    const filters= ['Name/URL', 'Version', 'Architecture'];
+    const versions = {};
+    const arches = {};
+
+    const context = useContext(ContentListContext)
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListFilters.tsx:12]: destructure where able.

File: src/components/ContentListTable/ContentListContext.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+  const [filterType, setFilterType] = useState<string>('Name/URL');
+  const [searchQuery, setSearchQuery] = useState('');
+  const [selectedVersions, setSelectedVersions] = useState<string[]>([]);
+  const [selectedArches, setSelectedArches] = useState<string[]>([]);
+
+  const filters = {
+    searchQuery, selectedVersions, selectedArches
+  }
+  const debouncedFilters = useDebounce(filters);
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListContext.tsx:52]: const {searchQuery: debouncedSearchQuery, selectedVersions: debouncedSelectedVersions, etc... } = useDebounce(filters)

File: src/services/useDebouce.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { useEffect, useState } from 'react'
```
Comment (Andrewgdewar) [inline src/services/useDebouce.tsx:1]: Change the name of this file (typo)

File: src/components/ContentListTable/ContentListTable.tsx (hunk 6 above)
Comment (rverdile) [reply src/components/ContentListTable/ContentListTable.tsx:70]: I'm a little unclear on what you mean. Are you saying to move this function to context and just call it in ContentListTable?

Comment (Andrewgdewar) [reply src/components/ContentListTable/ContentListTable.tsx:70]: No, I mean as you only consume these variables here: 
debouncedSearchQuery,
    debouncedSelectedVersions,
    debouncedSelectedArches, 
 
 .. I would suggest placing your function inside context, and have the "Labels" used in this component. 
 So the labels would be passed down by context.

Comment (rverdile) [reply src/components/ContentListTable/ContentListTable.tsx]: good idea! thanks

**Comment by rverdile** [comment]
I think I've addressed all the comments now

File: src/components/ContentListTable/ContentListChips.tsx [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import React, { useContext } from 'react';
+import { Button, Chip, ChipGroup } from '@patternfly/react-core';
+import { ContentListContext } from './ContentListContext';
+
+const ContentListChips = () => {
+
+    const context = useContext(ContentListContext)
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListChips.tsx:7]: destructure

File: src/components/ContentListTable/ContentListChips.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+    const clearFilters = () => {
+        context.setSelectedVersions([])
+        context.setSelectedArches([])
+        context.setSearchQuery('')
+    }
+
+    return (
+        <>
+            <ChipGroup categoryName={'Version'}>
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListChips.tsx:23]: Where there are strings as variables you can do ='Version'

File: src/components/ContentListTable/ContentListChips.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            </ChipGroup>
+                {(context.searchQuery !== '') && <ChipGroup categoryName={'Name/URL'}>
+                    <Chip key={'search_chip'} onClick={() => context.setSearchQuery('')}>
+                        {context.searchQuery}
+                    </Chip>
+            </ChipGroup>}
+            {(context.debouncedSearchQuery !== ''
+                && context.searchQuery !== ''
+                || context.selectedVersions.length !== 0
+                || context.selectedArches.length !== 0
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListChips.tsx:45]: you can use the double bang here to say not 0, as javascript considers 0 a falsey value. So: 
!!context.selectedArches?.length

File: src/components/ContentListTable/ContentListContext.tsx [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        const found = namesLabels.find(v => v.name === name )
+        if(found) {
+          labels.push(found.label)
+        }
+      }
+    );
+    return labels
+  };
+
+  const labels = {
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListContext.tsx:91]: 👍

File: src/components/ContentListTable/ContentListContext.tsx [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+  } = useDebounce(filters)
+
+  const {
+    isLoading,
+    error,
+    isError,
+    data = { distribution_arches: [], distribution_versions: [] },
+  } = useContentRepoParamsQuery();
+
+  const getLabels = (type: string, names: Array<string>) => {
```
[the remark this answers is not in the document]
Comment (Andrewgdewar) [reply src/components/ContentListTable/ContentListContext.tsx:78]: If you do go the route of using a useCallback, I'd suggest returning an object containing both of the labels you need.

File: src/components/ContentListTable/ContentListContext.tsx [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+  };
+
+  const labels = {
+    versions: getLabels('version', debouncedSelectedVersions),
+    arches: getLabels('arch', debouncedSelectedArches)
+  }
+
+  if (isError) throw error
+
+  if (isLoading) return <></>;
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListContext.tsx:98]: I have the loading logic built out on my branch, when this is merged I'll likely need to do something here.
```suggestion
  if (isLoading) return <></>; // TODO: Implement loader to match the top-level loader here.
```

File: src/components/ContentListTable/ContentListFilters.tsx [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+import DropdownSelect from '../DropdownSelect/DropdownSelect';
+import { FilterIcon } from '@patternfly/react-icons';
+import { ContentListContext } from './ContentListContext';
+
+const ContentListFilters = () => {
+    const filters= ['Name/URL', 'Version', 'Architecture'];
+    const versions = {};
+    const arches = {};
+
+    const {
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListFilters.tsx:12]: 👍

File: src/components/ContentListTable/ContentListFilters.tsx [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        setSearchQuery,
+        filterType,
+        setFilterType,
+        selectedVersions,
+        setSelectedVersions,
+        selectedArches,
+        setSelectedArches,
+    } = useContext(ContentListContext)
+
+    repoParams['distribution_arches'].forEach(arch => arches[arch.name] = arch.label );
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListFilters.tsx:23]: you'll want to to throw this into a useEffect with dependencies preventing it running every render.

File: src/components/ContentListTable/ContentListTable.tsx [hunk 20]
```diff
@@ -51,6 +46,11 @@ const ContentListTable = () => {
   const [page, setPage] = useState(1);
   const [perPage, setPerPage] = useState(storedPerPage);
 
+  const {
+    debouncedSearchQuery,
+    labels
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:51]:
```suggestion
   filterData
```

Comment (Andrewgdewar) [reply src/components/ContentListTable/ContentListTable.tsx:51]: Screen Shot 2022-07-19 at 10 04 45 AM

Could we change this to filterData? As this is only used for the API we can keep the idea of it similar to within the API.

File: src/components/ContentListTable/ContentListChips.tsx (hunk 13 above)
Comment (Andrewgdewar) [reply src/components/ContentListTable/ContentListChips.tsx:23]: Don't worry about this, I added a lint rule to fix/cover it.

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/14676876

File: src/components/ContentListTable/ContentListFilters.tsx [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    repoParams['distribution_versions'].forEach(version => versions[version.name] = version.label);
+
+    const onFilterChange = (): ReactElement => {
```
[the remark this answers is not in the document]
Comment (rverdile) [reply src/components/ContentListTable/ContentListFilters.tsx:32]: trying to do this, but changing this to a component causes a error "Can't perform a React state update on an unmounted component"

[the remark this answers is not in the document]
Comment (Andrewgdewar) [reply src/components/ContentListTable/ContentListFilters.tsx:32]: hmm.. 
Will look with you.

[the remark this answers is not in the document]
Comment (rverdile) [reply src/components/ContentListTable/ContentListFilters.tsx]: just renamed instead like discussed

**Comment by Andrewgdewar** [comment]
Screen Shot 2022-07-22 at 12 49 51 PM
Screen Shot 2022-07-22 at 12 50 36 PM

**Comment by Andrewgdewar** [comment]
Screen Shot 2022-07-22 at 12 52 14 PM
Screen Shot 2022-07-22 at 12 52 34 PM

File: src/components/ContentListTable/ContentListTable.tsx [hunk 22]
```diff
@@ -8,35 +8,30 @@ import {
...
 
 const useStyles = createUseStyles({
   actionContainer: {
     backgroundColor: global_BackgroundColor_100.value,
     justifyContent: 'space-between',
     padding: '16px 24px',
   },
+  chipsContainer: {
+    backgroundColor: global_BackgroundColor_100.value,
+    justifyContent: 'justifyContentFlexStart',
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:32]: does this work? I would think that the correct format would be flex-start here, as "justifyContentFlexStart" is specific to patternfly.

File: src/components/DropdownSelect/DropdownSelect.tsx [hunk 23]
```diff
@@ -0,0 +1,55 @@
+import { Select, SelectOption, SelectVariant } from '@patternfly/react-core';
+import React, { Dispatch, SetStateAction, useState } from 'react';
+
+type DropdownSelectProps = {
+    options: Array<string>,
+    variant: SelectVariant.single | SelectVariant.checkbox
+    selectedProp: string | Array<string>,
+    setSelected: Dispatch<SetStateAction<any>>, // TODO type this
```
Comment (Andrewgdewar) [inline src/components/DropdownSelect/DropdownSelect.tsx:8]: Your type will likely look something like this: (type: 'Name/URL' | 'Version' | 'Architecture') => void;

Comment (rverdile) [reply src/components/DropdownSelect/DropdownSelect.tsx:8]: I'm not sure this works. The issue here is I wanted to be able to set a string (or filter in this case) or Array<string> (for versions/arches). The "single" variant would be setting a string, the "checkbox" variant would be setting an array.

File: src/components/ContentListTable/ContentListTable.tsx (hunk 22 above)
Comment (rverdile) [reply src/components/ContentListTable/ContentListTable.tsx:32]: i changed it to 'flex-start' and it still looks okay, so I'll leave it like that. I don't remember where "justifyContentFlexStart" came from.

File: src/components/DropdownSelect/DropdownSelect.tsx (hunk 23 above)
Comment (Andrewgdewar) [reply src/components/DropdownSelect/DropdownSelect.tsx:8]: Then just add both as the types with an "|" or operator?

Comment (rverdile) [reply src/components/DropdownSelect/DropdownSelect.tsx]: It doesn't seem to be as simple as that. I pushed up what I tried, but I can't seem to find a way to type this that doesn't complain.

`setSelected: Dispatch<SetStateAction<Array<string>>> | Dispatch<SetStateAction<Filters>>`

Comment (Andrewgdewar) [reply src/components/DropdownSelect/DropdownSelect.tsx]: setSelected: Dispatch<SetStateAction<Filters | Array<string>>>;

Comment (Andrewgdewar) [reply src/components/DropdownSelect/DropdownSelect.tsx]: After further attempts at fixing this, I am now onboard for the "any" method within the DropdownSelect. 😅

File: src/components/ContentListTable/ContentListTable.tsx [hunk 24]
```diff
@@ -131,6 +136,11 @@ const ContentListTable = () => {
           />
         </FlexItem>
       </Flex>
+      <Hide hide={!selectedVersions.length && !selectedArches.length && searchQuery == ''}>
```
Comment (Andrewgdewar) [inline src/components/ContentListTable/ContentListTable.tsx:139]:
```suggestion
      <Hide hide={!showListChips}>
```

**Review by Andrewgdewar — APPROVED** [review]
Looking good, last comment was a bit #nit feel free to merge as is or with an update.

**Pull request merged**